File: repos_estudiantes_ALVAREZ_tercera_entrega/AJATA_CANAVIRI_NICOL_MAYRA/Persistencia_de_Objetos/Ejercicio3/Java/ArchivoCliente.java
package Persistencia_de_Objetos.Ejercicio3.Java;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class ArchivoCliente {
    private String nomArchivo;

    public ArchivoCliente(String nomArchivo) {
        this.nomArchivo = nomArchivo;
        crearArchivo(); 
    }

    private void crearArchivo() {
        File file = new File(nomArchivo);
        if (!file.exists()) {
            try {
                if (file.createNewFile()) {
                    System.out.println("Archivo '" + nomArchivo + "' creado exitosamente.");
                } else {
                    System.err.println("No se pudo crear el archivo '" + nomArchivo + "'.");
                }
            } catch (IOException e) {
                System.err.println("Error al crear el archivo '" + nomArchivo + "': " + e.getMessage());
            }
        } else {
            System.out.println("Archivo '" + nomArchivo + "' ya existe.");
        }
    }

    public void guardaCliente(Cliente c) {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(nomArchivo, true))) { // true para append
            writer.write(c.toString());
            writer.newLine(); 
            System.out.println("Cliente con ID " + c.getId() + " guardado exitosamente en '" + nomArchivo + "'.");
        } catch (IOException error) {
            System.err.println("Error al guardar cliente en '" + nomArchivo + "': " + error.getMessage());
        }
    }

    private List<Cliente> cargarTodosClientes() {
        List<Cliente> clientes = new ArrayList<>();
        try (BufferedReader reader = new BufferedReader(new FileReader(nomArchivo))) {
            String linea;
            while ((linea = reader.readLine()) != null) {
                Cliente cliente = Cliente.fromString(linea);
                if (cliente != null) {
                    clientes.add(cliente);
                }
            }
        } catch (IOException e) {
            System.err.println("Error al leer el archivo '" + nomArchivo + "': " + e.getMessage());
        }
        return clientes;
    }

    // b) 
    public Cliente buscarCliente(int idBuscar) {
        System.out.println("Buscando cliente con ID: " + idBuscar);
        List<Cliente> clientes = cargarTodosClientes();
        for (Cliente cliente : clientes) {
            if (cliente.getId() == idBuscar) {
                return cliente;
            }
        }
        return null;
    }

    // c) 
    public Cliente buscarCelularCliente(int telefonoBuscar) {
        System.out.println("Buscando cliente con numero de telefono: " + telefonoBuscar);
        List<Cliente> clientes = cargarTodosClientes();
        for (Cliente cliente : clientes) {
            if (cliente.getTelefono() == telefonoBuscar) {
                return cliente;
            }
        }
        return null; 
    }

    public void limpiarArchivo() {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(nomArchivo))) {
            writer.write(""); 
            System.out.println("Contenido del archivo '" + nomArchivo + "' limpiado.");
        } catch (IOException e) {
            System.err.println("Error al limpiar el archivo '" + nomArchivo + "': " + e.getMessage());
        }
    }
}

File: repos_estudiantes_ALVAREZ_tercera_entrega/AJATA_CANAVIRI_NICOL_MAYRA/Persistencia_de_Objetos/Ejercicio3/Python/ArchivoCliente.py

import os
from typing import List, Optional
from Cliente import Cliente

class ArchivoCliente:
    
    def __init__(self, nom_archivo: str):
        self.nom_archivo = nom_archivo
        self.crear_archivo()

    def crear_archivo(self):
        
        if not os.path.exists(self.nom_archivo):
            try:
                with open(self.nom_archivo, 'w') as f:
                    pass  
                print(f"Archivo '{self.nom_archivo}' creado exitosamente.")
            except IOError as e:
                print(f"Error al crear el archivo '{self.nom_archivo}': {e}")
        else:
            print(f"Archivo '{self.nom_archivo}' ya existe.")

    def guarda_cliente(self, c: Cliente):
        
        try:
            with open(self.nom_archivo, 'a') as f: 
                f.write(str(c) + '\n')
            print(f"Cliente con ID {c.id} guardado exitosamente en '{self.nom_archivo}'.")
        except IOError as error:
            print(f"Error al guardar cliente en '{self.nom_archivo}': {error}")

    def _cargar_todos_clientes(self) -> List[Cliente]:
       
        clientes = []
        try:
            with open(self.nom_archivo, 'r') as f:
                for linea in f:
                    cliente = Cliente.from_string(linea)
                    if cliente:
                        clientes.append(cliente)
            return clientes
        except FileNotFoundError:
            print(f"El archivo '{self.nom_archivo}' no se encontro. Retornando lista vacia.")
            return []
        except IOError as e:
            print(f"Error al leer el archivo '{self.nom_archivo}': {e}")
            return []

    # b) 
    def buscar_cliente(self, id_buscar: int) -> Optional[Cliente]:
        
        print(f"Buscando cliente con ID: {id_buscar}")
        clientes = self._cargar_todos_clientes()
        for cliente in clientes:
            if cliente.get_id() == id_buscar:
                return cliente
        return None

    # c) 
    def buscar_celular_cliente(self, telefono_buscar: int) -> Optional[Cliente]:
        
        print(f"Buscando cliente con numero de telefono: {telefono_buscar}")
        clientes = self._cargar_todos_clientes()
        for cliente in clientes:
            if cliente.get_telefono() == telefono_buscar:
                return cliente
        return None

    def limpiar_archivo(self):
       
        try:
            with open(self.nom_archivo, 'w') as f:
                f.write('')
            print(f"Contenido del archivo '{self.nom_archivo}' limpiado.")
        except IOError as e:
            print(f"Error al limpiar el archivo '{self.nom_archivo}': {e}")